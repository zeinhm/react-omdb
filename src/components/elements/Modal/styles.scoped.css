.modal-component {
  background-color: #fff;
  box-shadow: -0.6875rem 0.5625rem 1.3125rem rgba(173, 182, 217, 0.170884);
  border-radius: 0.3125rem;
  margin: auto;
  max-width: 34.125rem;
  padding: 2rem 2.8125rem;
  position: relative;
  top: 50%;
  transform: translateY(-50%);
}