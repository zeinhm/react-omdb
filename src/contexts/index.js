import React, { useState } from 'react';
import PropTypes from 'prop-types';

export const AppContext = React.createContext(null);
export const AlertContext = React.createContext(null);
export const ModalContext = React.createContext(null);

export default function AppContextProvider({ children }) {
  const [alert, setAlert] = useState(null);
  const [breadcrumbs, setBreadcrumbs] = useState([]);
  const [content, setContent] = useState(null);
  const appValue = {
    breadcrumbs,
    setBreadcrumbs,
  };
  const alertValue = {
    alert,
    setAlert,
  };
  const modalValue = {
    content,
    setContent,
  };

  return (
    <AppContext.Provider value={appValue}>
      <AlertContext.Provider value={alertValue}>
        <ModalContext.Provider value={modalValue}>
          {children}
        </ModalContext.Provider>
      </AlertContext.Provider>
    </AppContext.Provider>
  );
}

AppContextProvider.defaultProps = {
  children: null,
};

AppContextProvider.propTypes = {
  children: PropTypes.node,
};
