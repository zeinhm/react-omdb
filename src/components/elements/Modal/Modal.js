import React, { useContext, useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { ModalContext } from '../../../contexts';
import './styles.scoped.css';

export default function Modal(props) {
  const { content, setContent } = useContext(ModalContext);
  const { open, className, children, onClose } = props;
  const [modalContent, setModalContent] = useState(content);
  
  useEffect(() => {
    if (open) setContent(<Content children={children} className={className} />);
    else setContent(null);
  }, [open]);

  useEffect(() => {
    if (modalContent && !content) {
      onClose();
      setModalContent(content);
    } else {
      setModalContent(content);
    }
  }, [children]);

  return null;
}

Modal.defaultProps = {
  children: null,
  className: '',
  onClose: () => { },
  open: false,
};

Modal.propTypes = {
  children: PropTypes.node,
  className: PropTypes.string,
  onClose: PropTypes.func,
  open: PropTypes.bool,
};

export function Content({ className, children }) {
  const customClass = ['modal-component', className].filter(Boolean).join(' ');
  return (
    <section className={customClass}>
      {children}
    </section>
  );
}

Content.defaultProps = {
  children: null,
  className: '',
};

Content.propTypes = {
  children: PropTypes.node,
  className: PropTypes.string,
};
