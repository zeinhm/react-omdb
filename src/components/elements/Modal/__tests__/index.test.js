import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import Modal, { Content } from '../Modal';
import { ModalContext } from '../../../../contexts';

jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useContext: v => v,
  useEffect: cb => cb(),
  useState: v => [v === 'tes' ? v : !v, jest.fn()],
}));

jest.mock('../../../../contexts', () => ({
  ModalContext: {
    content: '',
    setContent: jest.fn(),
  },
}));

describe('src/components/elements/Modal', () => {
  test('render', () => {
    Modal.defaultProps.onClose();
    Modal.defaultProps.onClose = jest.fn();
    Modal(Modal.defaultProps);
    expect(ModalContext.setContent).toHaveBeenNthCalledWith(1, null);
    expect(Modal.defaultProps.onClose).toHaveBeenCalled();

    const openProps = { ...Modal.defaultProps, open: true };
    const { className, children } = Modal.defaultProps;
    const contentProps = { children, className, };
    ModalContext.content = 'contnpm ent';
    Modal(openProps);
    expect(ModalContext.setContent).toHaveBeenNthCalledWith(2, <Content {...contentProps} />);

    ModalContext.content = 'tes';
    Modal(Modal.defaultProps);
  });

  test('Content', () => {
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<Content />);
    expect(tree).toMatchSnapshot();
  });
});
