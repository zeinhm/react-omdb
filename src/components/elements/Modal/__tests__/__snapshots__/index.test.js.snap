// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/components/elements/Modal Content 1`] = `
<section
  className="modal-component"
/>
`;
